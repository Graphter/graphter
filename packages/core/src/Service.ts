import {ListResult} from "./ListResult";
import {GetResult} from "./GetResult";
import {SaveResult} from "./SaveResult";

export interface Service {
  list: (modelId: string, skip: number, take: number) => Promise<ListResult>,
  get: (modelId: string, instanceId: string | number) => Promise<GetResult>,
  save: (modelId: string, instance: any) => Promise<SaveResult>
}