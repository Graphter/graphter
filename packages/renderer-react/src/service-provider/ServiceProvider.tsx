import React, {createContext, useContext} from 'react';
import { NodeConfig, PathSegment, Service } from "@graphter/core";

export interface ServiceRegistration {
  id: string,
  service: Service
}

interface ServiceProviderProps {
  serviceRegistry: Array<ServiceRegistration | Array<ServiceRegistration>>
  children: any
}

const Context = createContext<Map<PathSegment, Service> | null>(null);

let serviceMap: Map<PathSegment, Service>

export function getService(id: string): Service {
  if(!serviceMap) throw new Error(`Couldn't find the API service registry. Make sure you've declared a <ServiceProvider /> and passed it a valid service.`)
  const service = serviceMap.get(id)
  if(!service) throw new Error(`Missing a service to handle '${id}' data`)
  return service
}

export function useService(id: string): Service {
  const serviceMap = useContext(Context)
  if(!serviceMap) throw new Error(`Couldn't find the API service registry. Make sure you've declared a <ServiceProvider /> and passed it a valid service.`)
  const service = serviceMap.get(id)
  if(!service) throw new Error(`Missing a service to handle '${id}' data`)
  return service
}

export default function ServiceProvider({ serviceRegistry, children }: ServiceProviderProps){
  serviceMap = serviceRegistry.reduce((a, c) => {
    if(Array.isArray(c)) c.forEach(registration => a.set(registration.id, registration.service))
    else a.set(c.id, c.service)
    return a
  }, new Map<PathSegment, Service>())
  return (
    <Context.Provider value={serviceMap}>
      {children}
    </Context.Provider>
  );
}