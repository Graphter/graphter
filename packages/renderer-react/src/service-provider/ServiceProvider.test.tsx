import React from 'react';
import ServiceProvider, {useService} from "./ServiceProvider";
import {Service} from "./Service";
import { act, render, within } from '@testing-library/react';
import { screen } from '@testing-library/dom'

describe(`<DataProvider />`, () => {
  let mockService: Service;

  function MockConsumer() {
    const service = useService();
    service.get('model-id', 'instance-id' );
    return null;
  }

  beforeEach(() => {
    mockService = {
      list: jest.fn().mockResolvedValue({
        items: [],
        count: 10,
        skip: 0,
        take: 10
      }),
      get: jest.fn().mockResolvedValue({
        item: {},
        version: 1,
        created: new Date(),
        updated: new Date(),
        deleted: false
      }),
      save: jest.fn().mockResolvedValue({
        item: {},
        version: 2,
        updated: new Date()
      }),
    };
  });

  it(`should render children`, () => {
    const { container } = render(
      <ServiceProvider service={mockService}>
        <div data-testid='some-child-component'>Some child component</div>
      </ServiceProvider>
    );
    const childComponent = screen.queryByTestId('some-child-component');
    expect(childComponent).not.toBeNull();
  });

  describe(' when using useService', () => {

    it(`should error when no provider is declared`, () => {
      expect(() => (
        render(<MockConsumer />)
      )).toThrowErrorMatchingSnapshot();
    });

    it(`should return the service passed to the provider`, () => {
      render(
        <ServiceProvider service={mockService}>
          <MockConsumer/>
        </ServiceProvider>
      );
      expect(mockService.get).toHaveBeenCalledWith('model-id', 'instance-id')
    });

  });

});


export {};