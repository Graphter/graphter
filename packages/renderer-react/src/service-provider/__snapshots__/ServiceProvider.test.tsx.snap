// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataProvider />  when using useService should error when no provider is declared 1`] = `"Couldn't find API service. Make sure you've declared a DataProvider and passed it a valid service."`;
